File: 2021/day-2/part-2/main.clj
(defn read-input
  [filename]
  (->> filename
       slurp
       clojure.string/split-lines))

(defn filter-direction
  [item]
  (first (clojure.string/split item #" ")))

(defn filter-unit
  [item]
  (Integer/parseInt (last (clojure.string/split item #" "))))

(defn calculate
  [input]
  (let [horizontal (atom 0)
        depth (atom 0)
        aim (atom 0)]
    (doseq [item input]
      (let [direction (filter-direction item)
            unit (filter-unit item)]
        (cond
          (= direction "up") (swap! aim - unit)
          (= direction "down") (swap! aim + unit)
          (= direction "forward") (do
                                    (swap! horizontal + unit)
                                    (swap! depth + (* @aim unit))))))
    (* @depth @horizontal)))

(defn main
  []
  (let [input (->> "input.txt"
                   read-input)]
    (->> input
         calculate)))

(print (main))

File: 2021/day-2/part-2/main.jl
function main()
    local horizontal = 0
    local depth = 0
    local aim = 0
    local content = readlines("input.txt")
    for line in content
        direction, unit = split(line, " ")
        unit = parse.(Int, unit)
        if direction == "forward"
            horizontal += unit
            depth += aim * unit
        elseif direction == "up"
            aim -= unit
        elseif direction == "down"
            aim += unit
        end
    end
    println(horizontal * depth)
end

main()